### ISSUE TITLE
SSL_get0_alpn_selected: *data == NULL while len != 0 leading to potential segmentation fault.

### ISSUE THREAD

**Issue by thican** [libressl-portable/portable#614]
Hello,

EDIT: Adds more information.

I reported an issue on cURL’s issues tracker about a segmentation fault with their CLI:  https://github.com/curl/curl/issues/5926
After more investigation, we noticed this issue comes from the LibreSSL `SSL_get0_alpn_selected` function, and according to its documentation[1], we found an undocumented behaviour where `*data == NULL` while `len != 0` (hence the title).

1: https://www.openssl.org/docs/man1.0.2/man3/SSL_get0_alpn_selected.html

To avoid losing investigation, here the full backtrace of cURL after the call of this function:
~~~
% gdb -c /tmp/core_curl_1599393760_11527 "$(which curl)"
GNU gdb (Gentoo 9.2 vanilla) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/curl...
Reading symbols from /usr/lib64/debug//usr/bin/curl.debug...
[New LWP 11527]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `curl --include --head https://blog.cacert.org/'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __memcmp_sse4_1 () at ../sysdeps/x86_64/multiarch/memcmp-sse4.S:802
802             mov     -8(%rsi), %rcx
gdb-peda$ bt full
#0  __memcmp_sse4_1 () at ../sysdeps/x86_64/multiarch/memcmp-sse4.S:802
No locals.
#1  0x00007f3623dee01e in ossl_connect_step2 (conn=conn@entry=0x55bc32216af0, sockindex=sockindex@entry=0x0)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/vtls/openssl.c:3346
        neg_protocol = 0x0
        len = 0x8
        data = 0x55bc32218220
        err = 0x1
        connssl = 0x55bc32217000
        certverifyresult = 0x55bc32218ca8
        backend = 0x55bc32218070
#2  0x00007f3623dee18d in ossl_connect_common (conn=0x55bc32216af0, sockindex=0x0, nonblocking=nonblocking@entry=0x1, 
    done=0x7fff4711de68) at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/vtls/openssl.c:4037
        timeout_ms = <optimized out>
        result = <optimized out>
        data = 0x55bc32218220
        connssl = 0x55bc32217000
        sockfd = 0x5
        what = <optimized out>
#3  0x00007f3623dee31e in Curl_ossl_connect_nonblocking (conn=<optimized out>, sockindex=<optimized out>, done=<optimized out>)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/vtls/openssl.c:4071
No locals.
#4  0x00007f3623def06a in Curl_ssl_connect_nonblocking (conn=conn@entry=0x55bc32216af0, sockindex=sockindex@entry=0x0, 
    done=0x7fff4711de68) at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/vtls/vtls.c:331
        result = <optimized out>
#5  0x00007f3623db6ce5 in https_connecting (conn=0x55bc32216af0, done=<optimized out>)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/http.c:1495
        result = <optimized out>
#6  0x00007f3623dc808e in protocol_connecting (conn=<optimized out>, done=done@entry=0x7fff4711de68)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/multi.c:1459
        result = CURLE_OK
#7  0x00007f3623dcadf6 in multi_runsingle (multi=multi@entry=0x55bc32216870, now=..., data=data@entry=0x55bc32218220)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/multi.c:1885
        stream_error = 0x0
        msg = <optimized out>
        connected = 0x0
        async = 0x0
        protocol_connected = 0x0
        dophase_done = 0x0
        done = 0x0
        rc = CURLM_OK
        result = CURLE_OK
        timeout_ms = <optimized out>
        recv_timeout_ms = <optimized out>
        send_timeout_ms = <optimized out>
        control = 0x7fff
#8  0x00007f3623dcbaa5 in curl_multi_perform (multi=multi@entry=0x55bc32216870, running_handles=running_handles@entry=0x7fff4711dfe0)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/multi.c:2420
        result = <optimized out>
        pipe_st = {
          old_pipe_act = {
            __sigaction_handler = {
              sa_handler = 0x1,
              sa_sigaction = 0x1
            },
            sa_mask = {
              __val = {0x1000, 0x7f3623d664d0, 0x1000, 0x7f3623d664d0, 0x1000, 0x7f3623d664d0, 0x1000, 0x7f3623d664d0, 0x1000, 
                0x7f3623d664d0, 0x7fff47148808, 0x1, 0x7fff4711de20, 0x7fff4711ddf0, 0x7f3623c5d583, 0x7fff4711deb0}
            },
            sa_flags = 0x14000000,
            sa_restorer = 0x7f3623d664d0 <__restore_rt>
          },
          no_signal = 0x0
        }
        data = 0x55bc32218220
        returncode = CURLM_OK
        t = 0x100000003
        now = <optimized out>
#9  0x00007f3623da4ae9 in easy_transfer (multi=multi@entry=0x55bc32216870)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/easy.c:603
        still_running = 0x0
        done = 0x0
        mcode = CURLM_OK
        result = CURLE_OK
#10 0x00007f3623da4c62 in easy_perform (data=0x55bc32218220, events=events@entry=0x0)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/easy.c:702
        multi = 0x55bc32216870
        mcode = CURLM_OK
        result = CURLE_OK
        pipe_st = {
          old_pipe_act = {
            __sigaction_handler = {
              sa_handler = 0x1,
              sa_sigaction = 0x1
            },
            sa_mask = {
              __val = {0x1000, 0x1, 0x700000031, 0x90, 0x8595d27caa332900, 0x7fff4711df70, 0x7f3623c87511, 0x7, 0x7fff47148808, 0x1, 
                0x7fff4711df80, 0x7fff4711df50, 0x7f3623c5d583, 0x55bc32218220, 0x8, 0x7fff4711dfa0}
            },
            sa_flags = 0x14000000,
            sa_restorer = 0x7f3623d664d0 <__restore_rt>
          },
          no_signal = 0x0
        }
#11 0x00007f3623da4fd6 in curl_easy_perform (data=<optimized out>) at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/lib/easy.c:712
No locals.
#12 0x000055bc30ca6403 in serial_transfers (global=0x7fff4711e200, share=0x55bc32213370)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/src/tool_operate.c:2291
        retry = 0x0
        bailout = 0x0
        returncode = CURLE_OK
        result = CURLE_OK
        per = 0x55bc32217bf0
        added = 0x1
#13 0x000055bc30ca64fa in run_all_transfers (global=global@entry=0x7fff4711e200, share=share@entry=0x55bc32213370, result=CURLE_OK)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/src/tool_operate.c:2467
        orig_noprogress = 0x0
        orig_isatty = 0x0
        per = <optimized out>
#14 0x000055bc30ca676e in operate (global=0x7fff4711e200, argc=<optimized out>, argv=<optimized out>)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/src/tool_operate.c:2582
        count = 0x1
        operation = 0x0
        share = 0x55bc32213370
        res = <optimized out>
        result = <optimized out>
        first_arg = <optimized out>
#15 0x000055bc30c9f84a in main (argc=argc@entry=0x4, argv=argv@entry=0x7fff4711e398)
    at /usr/src/debug/net-misc/curl-7.72.0/curl-7.72.0/src/tool_main.c:323
        result = CURLE_OK
        global = {
          showerror = 0xffffffff,
          mute = 0x0,
          noprogress = 0x1,
          isatty = 0x1,
          errors = 0x7f3623d4e420 <_IO_2_1_stderr_>,
          errors_fopened = 0x0,
          trace_dump = 0x0,
          trace_stream = 0x0,
          trace_fopened = 0x0,
          tracetype = TRACE_NONE,
          tracetime = 0x0,
          progressmode = 0x0,
          libcurl = 0x0,
          fail_early = 0x0,
          styled_output = 0x1,
          parallel = 0x0,
          parallel_max = 0x32,
          parallel_connect = 0x0,
          first = 0x55bc321fb2d0,
          current = 0x55bc321fb2d0,
          last = 0x55bc321fb2d0
        }
#16 0x00007f3623bb9d23 in __libc_start_main (main=0x55bc30c9f7bf <main>, argc=0x4, argv=0x7fff4711e398, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fff4711e388) at ../csu/libc-start.c:308
        self = <optimized out>
        result = <optimized out>
        unwind_buf = {
          cancel_jmp_buf = {{
              jmp_buf = {0x55bc30cac440, 0xefdd52d67ee393ec, 0x0, 0x7fff4711e390, 0x0, 0x0, 0xefdd52d67d6393ec, 0xee4f9b8281a993ec},
              mask_was_saved = 0x0
            }},
          priv = {
            pad = {0x0, 0x0, 0x7f3623e50130, 0x7fff4711e380},
            data = {
              prev = 0x0,
              cleanup = 0x0,
              canceltype = 0x23e50130
            }
          }
        }
        not_first_call = <optimized out>
#17 0x000055bc30c9685a in _start ()
No symbol table info available.
~~~

Here the current state of the source code of cURL (7.72.0) where there is the call to the function `SSL_get0_alpn_selected` (line 3335) followed by the `memcmp` call (inside the conditional test starting line 3346), with `NULL` as parameter for `const void *s2`, therefore triggering the segfault.
https://github.com/curl/curl/blob/curl-7_72_0/lib/vtls/openssl.c#L3335

However, this issue doesn’t happen with cURL using OpenSSL as SSL/TLS backend.

Unfortunately, this is the only case I found about this issue.

Thanks for support.

**Comment by feliwir** [comment]
I'm getting this aswell

**Comment by kinichiro** [comment]
commit https://github.com/openbsd/src/commit/eef1bcf28fbc5bbca30f8b5490ec2bf15c0b9f72 will solve this issue.
I had tested with curl 7.72.0 and LibreSSL 3.1.4 with this patch.

**Comment by feliwir** [comment]
@kinichiro is there a new LibreSSL patch version planned for this?

**Comment by feliwir** [comment]
Can i please get any update? This issue is still blocking me..

**Comment by busterb** [comment]
@feliwir if there's an upstream errata patch issued, we'll cut a new stable portable build. Are you not able to apply the patch @kinichiro pointed out, or you just waiting for an official release?

**Comment by botovq** [comment]
On Mon, Oct 12, 2020 at 04:39:09AM -0700, Stephan Vedder wrote:
> Can i please get any update? This issue is still blocking me..

We will release the first stable version on the 3.2. branch within the
next week. This has this and many other things fixed I recommend that
you update to that version. We won't be releasing a reliability fix for
OpenBSD for this bug, so this won't be fixed on the 3.1 branch in the
foreseeable future.

If you want to patch 3.1 locally, you would need these two commits:

https://github.com/openbsd/src/commit/834e13ab1c912d405b11915ba98eaf628abeb8ee
https://github.com/openbsd/src/commit/eef1bcf28fbc5bbca30f8b5490ec2bf15c0b9f72

**Comment by feliwir** [comment]
> @feliwir if there's an upstream errata patch issued, we'll cut a new stable portable build. Are you not able to apply the patch @kinichiro pointed out, or you just waiting for an official release?

I need an official release, since that makes tracking down the dependency versions a lot easier in our QM-system. Applying the patches manually did work & solve the problem.

> On Mon, Oct 12, 2020 at 04:39:09AM -0700, Stephan Vedder wrote: Can i please get any update? This issue is still blocking me..
> We will release the first stable version on the 3.2. branch within the next week. This has this and many other things fixed I recommend that you update to that version. We won't be releasing a reliability fix for OpenBSD for this bug, so this won't be fixed on the 3.1 branch in the foreseeable future. If you want to patch 3.1 locally, you would need these two commits: [openbsd/src@834e13a](https://github.com/openbsd/src/commit/834e13ab1c912d405b11915ba98eaf628abeb8ee) [openbsd/src@eef1bcf](https://github.com/openbsd/src/commit/eef1bcf28fbc5bbca30f8b5490ec2bf15c0b9f72)

Thanks, glad to hear that!

**Comment by 4a6f656c** [comment]
This is fixed in LibreSSL 3.2.2, closing.
